raft/confchange/confchange.go

```diff
@@ -0,0 +1,375 @@
+// for correctness.
+func (c Changer) apply(cfg *tracker.Config, prs tracker.ProgressMap, ccs ...pb.ConfChange) error {
+	for _, cc := range ccs {
+		if cc.NodeID != 0 {
+			switch cc.Type {
+			case pb.ConfChangeAddNode:
+				c.makeVoter(cfg, prs, cc.NodeID)
+			case pb.ConfChangeAddLearnerNode:
+				c.makeLearner(cfg, prs, cc.NodeID)
+			case pb.ConfChangeRemoveNode:
+				c.remove(cfg, prs, cc.NodeID)
+			case pb.ConfChangeUpdateNode:
+			default:
+				return fmt.Errorf("unexpected conf type %d", cc.Type)
+			}
+		}
+	}
+	if len(cfg.Voters[0]) == 0 {
+		return errors.New("removed all voters")
+	}
+	return nil
+}
+
+// remove this peer as a voter or learner. If the peer is a voter on both sides
+// of the joint config, it will be removed only from the left.
+func (c Changer) remove(cfg *tracker.Config, prs tracker.ProgressMap, id uint64) {
+	if _, ok := prs[id]; !ok {
+		return
+	}
+
+	delete(cfg.Voters[0], id)
+	nilAwareDelete(&cfg.Learners, id)
+	nilAwareDelete(&cfg.LearnersNext, id)
+
+	// If the peer is still a voter on the right, keep the Progress.
+	if _, onRight := cfg.Voters[1][id]; !onRight {
+		delete(prs, id)
+	}
+}
+
+// makeVoter adds or promotes the given ID to be a voter on the left side of the
+// joint quorum. Match and next are only used if the voter was added (i.e. not
+// promoted).
+func (c Changer) makeVoter(cfg *tracker.Config, prs tracker.ProgressMap, id uint64) {
+	pr := prs[id]
+	if pr == nil {
+		c.initProgress(cfg, prs, id, false /* isLearner */)
+		return
+	}
+
+	pr.IsLearner = false
+	nilAwareDelete(&cfg.Learners, id)
+	nilAwareDelete(&cfg.LearnersNext, id)
+	cfg.Voters[0][id] = struct{}{}
+	return
+}
+
+// makeLearner makes the given ID a learner or stages it to be a learner once
+// an active joint configuration is exited.
+//
+// The former happens when the peer isn't already a voter on the *right hand
+// side* of the joint configuration: we either add a new learner or demote the
+// possibly existing voter on the *left hand side* of the joint configuration,
+// leaving the peer solely as a learner.
+//
+// The latter case occurs when the configuration is joint and the peer is a
+// voter on the right hand side. In that case, we cannot add the peer as a
+// learner at this moment because it is already a voter which can only be
+// removed by exiting the joint configuration first (which has to be done
+// explicitly by the caller only once they know it is safe to do). Instead, the
+// peer will be added as a "next learner", that is, it will become a learner the
+// moment the joint configuration is left.
+func (c Changer) makeLearner(cfg *tracker.Config, prs tracker.ProgressMap, id uint64) {
+	pr := prs[id]
+	if pr == nil {
+		c.initProgress(cfg, prs, id, true /* isLearner */)
+		return
+	}
+	if pr.IsLearner {
+		return
+	}
+	// Remove any existing voter on the left...
+	c.remove(cfg, prs, id)
+	// ... but save the Progress.
+	prs[id] = pr
+	if _, onRight := cfg.Voters[1][id]; onRight {
```

Comment:
This is subtle and relies on the contract you gave `LearnersNext`. It deserves a shoutout in a comment at a minimum.